### PR TITLE
Add --recursive option

### LINKED ISSUE #42: Include sub directories
I have a the following directory structure:
```
input
├── category1
│   ├── foo.svg
│   └── bar.svg
├── baz.svg
└── category2
    ├── foo.svg
    └── bar.svg
```

When calling `spreet` i only get a spritesheet including `baz.svg`. Instead I would like to include all files with the following ids: `category1/foo`, `category1/bar`, `baz`, `category2/foo`, `category2/bar`.

As far as I can tell this is not possible at the moment. Would you be open to a PR which adds this functionality?
If yes, should this this be disabled by default and behind an option (`--deep` or something)?

### PR DESCRIPTION
**Pull request by DerZade**
Add `--recursive` option and therefore fix #42.

---

This PR includes two breaking changes to the public API:
- added second parameter (`recurse`) to `spreet::fs::get_svg_input_paths`
- added second parameter (`base_path`) to `spreet::sprite::sprite_name`

I hope this is acceptable. If not, just let me know and I will add other functions instead of extending the existing ones. But this will result in some duplicate code 🙈

### CODE REVIEW

File: src/sprite.rs [hunk 1]
```diff
@@ -228,8 +228,18 @@ impl Spritesheet {
 }
 
 /// Returns the name (unique id within a spritesheet) taken from a file.
-pub fn sprite_name(path: &Path) -> String {
-    format!("{}", path.file_stem().unwrap().to_string_lossy())
+pub fn sprite_name(path: &Path, base_path: &Path) -> String {
```
Comment (flother) [inline src/sprite.rs:231]: Now it's doing something a little more complicated, I think this function's a perfect candidate for a unit test or two (I'm holding you to a higher standard than I hold myself, sorry!) Do you fancy adding some? One for when base_path has no impact, one when it does

File: src/sprite.rs [hunk 2]
```diff
@@ -228,8 +228,18 @@ impl Spritesheet {
 }
 
 /// Returns the name (unique id within a spritesheet) taken from a file.
```
Comment (flother) [inline src/sprite.rs:230]: Let's add an explanation of what `base_path` is used for

```suggestion
/// Returns the name (unique id within a spritesheet) taken from a file.
///
/// If `base_path` is given, it's used as prefix in the sprite name (minus any parents
/// shared with `path`).
```

**Review by flother — COMMENTED** [review]
This is great, I appreciate you taking the time to add this in. I have a few comments on minor things I'd like to see, but I'll be happy to merge this.

Other than the comments below, could you add a line to [`CHANGELOG.md`](https://github.com/flother/spreet/blob/master/CHANGELOG.md) to summarise the new flag?

File: README.md [hunk 3]
```diff
@@ -99,6 +99,7 @@ Options:
   -r, --ratio <RATIO>      Set the output pixel ratio [default: 1]
       --retina             Set the pixel ratio to 2 (equivalent to `--ratio=2`)
       --unique             Store only unique images in the spritesheet, and map them to multiple names
+      --recursive          Include images from subdirectories
```
[the remark this answers is not in the document]
Comment (DerZade) [reply README.md]: Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

File: src/fs.rs [hunk 4]
```diff
@@ -31,14 +31,25 @@ pub fn is_useful_input(entry: &DirEntry) -> bool {
 
 /// Returns a vector of file paths matching all SVGs within the given directory.
 ///
-/// This does not recurse into sub-directories; they are silently ignored. It ignores hidden files
-/// (files whose names begin with `.`) but it does follow symlinks.
-pub fn get_svg_input_paths(path: &Path) -> Vec<PathBuf> {
+/// It ignores hidden files (files whose names begin with `.`) but it does follow symlinks.
+pub fn get_svg_input_paths(path: &Path, recurse: bool) -> Vec<PathBuf> {
```
[the remark this answers is not in the document]
Comment (DerZade) [reply src/fs.rs]: Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

File: src/fs.rs [hunk 5]
```diff
@@ -31,14 +31,25 @@ pub fn is_useful_input(entry: &DirEntry) -> bool {
...
     read_dir(path)
         .unwrap()
         .flatten()
-        .filter(is_useful_input)
-        .map(|entry| entry.path())
+        .filter_map(|entry| {
+            let path_buf = entry.path();
+
+            if recurse && path_buf.is_dir() {
+                return Some(get_svg_input_paths(&path_buf.as_path(), recurse));
```
[the remark this answers is not in the document]
Comment (DerZade) [reply src/fs.rs]: Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

File: src/fs.rs [hunk 6]
```diff
@@ -31,14 +31,25 @@ pub fn is_useful_input(entry: &DirEntry) -> bool {
...
+            }
+
```
[the remark this answers is not in the document]
Comment (DerZade) [reply src/fs.rs]: Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

File: src/fs.rs [hunk 7]
```diff
@@ -31,14 +31,25 @@ pub fn is_useful_input(entry: &DirEntry) -> bool {
...
+            if is_useful_input(&entry) {
+                return Some(vec![path_buf]);
+            }
+
```
[the remark this answers is not in the document]
Comment (DerZade) [reply src/fs.rs]: Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

File: src/sprite.rs (hunk 2 above)
Comment (DerZade) [reply src/sprite.rs:230]: You're explanation starts with _"If `base_path` is given"_. There is not really a choice tho 😂 so I adjusted it a bit. 

Addressed in 7a6a857c1c630eeeefcf08bb992462a2f8afffa0

**Comment by DerZade** [comment]
I have addressed most of your comments and have added a line to the change log.

The only thing left to do is https://github.com/flother/spreet/pull/43#discussion_r1103876825, which I hope to get to tonight :)

**Comment by DerZade** [comment]
I added tests for `sprite_name` in 73fda65003dc95ce4a639e30c65a1ad57d4d6aa4 🙃

File: src/sprite.rs (hunk 1 above)
Comment (DerZade) [reply src/sprite.rs:234]: Addressed in 73fda65003dc95ce4a639e30c65a1ad57d4d6aa4

File: src/fs.rs [hunk 8]
```diff
@@ -31,14 +31,23 @@ pub fn is_useful_input(entry: &DirEntry) -> bool {
...
+        .filter_map(|entry| {
+            let path_buf = entry.path();
+
+            if recursive && path_buf.is_dir() {
+                return Some(get_svg_input_paths(path_buf.as_path(), recursive));
+            }
+            if is_useful_input(&entry) {
+                return Some(vec![path_buf]);
+            }
+            return None;
```
Comment (flother) [inline src/fs.rs:48]:
```suggestion
            if recursive && path_buf.is_dir() {
                Some(get_svg_input_paths(path_buf.as_path(), recursive))
            } else if is_useful_input(&entry) {
                Some(vec![path_buf])
            } else {
                None
            }
```

**Review by flother — APPROVED** [review]
Brilliant, thanks for taking the time to add this

**Comment by DerZade** [comment]
Thanks a lot for the most pleasant OSS experience I had in a long time :)

**Pull request merged**